{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This notebook contains step-by-step instructions for using Dropseqtools 1.13 (computational protocol 1.0.1)\n",
    "\n",
    "- The following notebook contains instructions for aligning reads generated from SeqWell/DropSeq protocols. Much of this is taken directly from the McCarroll Lab's [alignment cookbook](http://mccarrolllab.com/wp-content/uploads/2016/03/Drop-seqAlignmentCookbookv1.2Jan2016.pdf), with copied items being *italicized*. Here, we will run Dropseqtools on data from the first batch of data on mouse retina from the Macosko et al., 2015 paper.\n",
    "\n",
    "- We downloaded and  downsampled to a 100 Million Reads data from the first batch of cells in order to be able to process it interactively in a timely manner."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting started:\n",
    "\n",
    "We have downloaded all the published data and other files you might need for this and other sessions. In order to get access to these files you can create a softlink to the folder:\n",
    "\n",
    "```bash\n",
    "cd ~\n",
    "ln -s /oasis/tscc/scratch/cshl_2018/ ./\n",
    "```\n",
    "\n",
    "Finally, we will make another softlink in your home to your own scratch folder:\n",
    "\n",
    "```bash\n",
    "ln -s /oasis/tscc/scratch/ucsd-trainXX/ ./scratch\n",
    "```\n",
    "\n",
    "NOTE: Remember to use your own ucsd-train## (i.e. ucsd-train27)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Interactive Jobs:\n",
    "For dropseqtools we will be working with interactive jobs in order to visualize on real time the progress of each processing step as we go through the analysis. We have allocated Yeo Lab cluster resources for our use throughout the course. To request an interactive job:\n",
    "```bash\n",
    "qsub -I -l walltime=2:00:00 -q home-yeo -l nodes=1:ppn=2\n",
    "```\n",
    "\n",
    "keep in mind that you can adjust your ```walltime=H:MM:SS ``` as necessary."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Starting in the right enviroment:\n",
    "By this point you should have completed the [instructions for downloading and installing dropseqtools 0.13](add link) which means that you should be able to activate your dropseqtools environment\n",
    "\n",
    "```bash\n",
    "source activate dropseqtools\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preprocessing:\n",
    "*The raw reads from the sequencer must be converted into a Picard queryname sorted BAM file for\n",
    "each library in the sequencer run. Since there are many sequencers and pipelines available to do\n",
    "this, we leave this step to the user. For example, we use either Picard IlluminaBasecallsToSam\n",
    "(preceded by Picard ExtractIlluminaBarcodes for a library with sample barcodes);  or Illumina’s\n",
    "bcl2fastq followed by Picard FastqToSam.  Once you have an unmapped, queryname sorted BAM, you\n",
    "can follow this set of steps to align your raw reads and create a BAM file that is suitable to produce\n",
    "digital gene expression (DGE) results.*\n",
    "\n",
    "BAM files are typically known for storing alignment data, or reads that have been aligned to a reference. However, this format can also store reads that are unaligned (just like fastq.gz files), and there are exists some tools (including the ones referenced in this notebook) that use the BAM file format to store unaligned reads.\n",
    "\n",
    "So before proceeding, we will need to convert fastq.gz files to their BAM file equivalents (in this example, we will call our bam file ```my_unaligned_data.bam``` and will be used as an input into the next step:\n",
    "\n",
    "```bash\n",
    "picard FastqToSam \\\n",
    "F1=forward_reads.fastq \\\n",
    "F2=reverse_reads.fastq \\\n",
    "O=my_unaligned_data.bam \\\n",
    "SM=sample001 \\\n",
    "```\n",
    "\n",
    "**Note:** the backslash (```\\```) after each line indicates the continuation of a command on the next line."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alignment overview:\n",
    "- Unmapped BAM > aligned and tagged BAM\n",
    "- Tag cell barcodes\n",
    "- Tag molecular barcodes\n",
    "- Trim 5’ primer sequence\n",
    "- Trim 3’ polyA sequence\n",
    "- SAM > Fastq\n",
    "- STAR alignment\n",
    "- Sort STAR alignment in queryname order\n",
    "- Merge STAR alignment tagged SAM to recover cell/molecular barcodes\n",
    "- Add gene/exon and other annotation tags\n",
    "- DetectBeadSynthesisErrors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metadata:\n",
    "\n",
    "- **fasta**: *The reference sequence of the organism. Needed for most aligners.*\n",
    "- **dict**: A dictionary file as generated by Picard’s CreateSequenceDictionary. Needed for Picard\n",
    "Tools.*\n",
    "- **gtf**: *The principle file to determine the location of genomic features like genes, transcripts,\n",
    "and exons. Many other metadata files we use derive from this original file. We download our\n",
    "GTF files from ensembl, which has a handy description of the file format here. Ensembl has a\n",
    "huge number of prepared GTF files for a variety of organisms here.*\n",
    "- **refFlat**: *This file contains a subset of the the same information in the GTF file in a different\n",
    "format. Picard tools like the refFlat format, so we require this as well. To make life easy, we\n",
    "provide a program ConvertToRefFlat that can convert files from GTF format to refFlat for you.*\n",
    "- **genes.intervals**: *The genes from the GTF file in interval list format. This file is optional, and\n",
    "useful if you want to go back to your BAM later to see what gene(s) a read aligns to.*\n",
    "- **exons.intervals**: *The exons from the GTF file in interval list format. This file is optional, and\n",
    "useful if you want to go back to your BAM and view what exon(s) a read aligns to.*\n",
    "- **rRNA.intervals**: *The locations of ribosomal RNA in interval list format. This file is optional, but\n",
    "we find it useful to later assess how much of a dropseq library aligns to rRNA.*\n",
    "- **reduced.gtf**: *This file contains a subset of the information in the GTF file, but in a far more\n",
    "human readable format. This file is optional, but can be generated easily by the supplied\n",
    "ReduceGTF program that will take a GTF file as input.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TagBamWithReadsSequenceExtended (tag cells)\n",
    "\n",
    "*This Drop-seq program extracts bases from the cell/molecular barcode encoding read **(BARCODED_READ)**, and creates a new BAM tag with those bases on the genome read. By default, we use the BAM tag XM for molecular barcodes, and **XC** for cell barcodes, using the **TAG_NAME** parameter.*\n",
    "\n",
    "*This program is run once per barcode extraction to add a tag. On the first iteration, the cell barcode is\n",
    "extracted from bases **1-12**. This is controlled by the BASE_RANGE option.*\n",
    "\n",
    "Example Cell Barcode:\n",
    "\n",
    "```bash\n",
    "TagBamWithReadSequenceExtended \\\n",
    "INPUT=my_unaligned_data.bam \\\n",
    "OUTPUT=unaligned_tagged_Cell.bam \\\n",
    "SUMMARY=unaligned_tagged_Cellular.bam_summary.txt \\\n",
    "BASE_RANGE=1-12 \\\n",
    "BASE_QUALITY=10 \\\n",
    "BARCODED_READ=1 \\\n",
    "DISCARD_READ=False \\\n",
    "TAG_NAME=XC \\\n",
    "NUM_BASES_BELOW_QUALITY=1\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TagBamWithReadsSequenceExtended (tag umi)\n",
    "\n",
    "*On the second iteration, the molecular barcode is extracted from bases **13-20** of the barcode read. This program has an option to drop a read **(DISCARD_READ)**, which we use after both barcodes have been extracted, which makes\n",
    "the output BAM have unpaired reads with additional tags. Additionally, this program has a **BASE_QUALITY** option, which is the minimum base quality of all bases of the barcode being extracted. If more than **NUM_BASES_BELOW_QUALITY** bases falls below this quality, the read pair is discarded.*\n",
    "\n",
    "Example Molecular Barcode:\n",
    "\n",
    "```bash\n",
    "TagBamWithReadSequenceExtended \\\n",
    "INPUT=unaligned_tagged_Cell.bam \\\n",
    "OUTPUT=unaligned_tagged_CellMolecular.bam \\\n",
    "SUMMARY=unaligned_tagged_Molecular.bam_summary.txt \\\n",
    "BASE_RANGE=13-20 \\\n",
    "BASE_QUALITY=10 \\\n",
    "BARCODED_READ=1 \\\n",
    "DISCARD_READ=True \\\n",
    "TAG_NAME=XM \\\n",
    "NUM_BASES_BELOW_QUALITY=1\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FilterBam\n",
    "\n",
    "*This Drop-seq program is used to remove reads where the cell or molecular barcode has low quality\n",
    "bases. During the run of **TagBamWithReadSequenceExtended**, an **XQ** tag is added to each read to\n",
    "represent the number of bases that have quality scores below the **BASE_QUALITY** threshold. These\n",
    "reads are then removed from the BAM.*\n",
    "\n",
    "Example:\n",
    "```bash\n",
    "FilterBAM \\\n",
    "TAG_REJECT=XQ \\\n",
    "INPUT=unaligned_tagged_CellMolecular.bam \\\n",
    "OUTPUT=unaligned_tagged_filtered.bam \\\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TrimStartingSequence\n",
    "\n",
    "*This Drop-seq program is one of two sequence cleanup programs designed to trim away any extra\n",
    "sequence that might have snuck it’s way into the reads. In this case, we trim the SMART Adapter that\n",
    "can occur 5’ of the read. In our standard run, we look for at least 5 contiguous bases **(NUM_BASES)** of\n",
    "the SMART adapter **(SEQUENCE)** at the 5’ end of the read with no errors **(MISMATCHES)** , and hard\n",
    "clip those bases off the read.*\n",
    "\n",
    "\n",
    "Example:\n",
    "```bash\n",
    "TrimStartingSequence \\\n",
    "INPUT=unaligned_tagged_filtered.bam \\\n",
    "OUTPUT=unaligned_tagged_trimmed_smart.bam \\\n",
    "OUTPUT_SUMMARY=adapter_trimming_report.txt \\\n",
    "SEQUENCE=AAGCAGTGGTATCAACGCAGAGTGAATGGG \\\n",
    "MISMATCHES=0 \\\n",
    "NUM_BASES=5\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PolyATrimmer\n",
    "\n",
    "*This Drop-seq program is the second sequence cleanup program designed to trim away trailing polyA\n",
    "tails from reads. It searches for at least **6 (NUM_BASES)** contiguous A’s in the read with **0 (MISMATCHES)**, and hard clips the read to remove these bases and all bases 3’ of the polyA run.*\n",
    "\n",
    "Example:\n",
    "```bash\n",
    "PolyATrimmer \\\n",
    "INPUT=unaligned_tagged_trimmed_smart.bam \\\n",
    "OUTPUT=unaligned_mc_tagged_polyA_filtered.bam \\\n",
    "OUTPUT_SUMMARY=polyA_trimming_report.txt \\\n",
    "MISMATCHES=0 \\\n",
    "NUM_BASES=6\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SamToFastq\n",
    "\n",
    "*Now that your data has had the cell and molecular barcodes extracted, the reads have been cleaned\n",
    "of SMARTSeq primer and polyA tails, and the data is now unpaired reads, it’s time to align. To do this,\n",
    "we extract the FASTQ files using Picard’s **SamToFastq** program.*\n",
    "\n",
    "```bash\n",
    "picard SamToFastq \\\n",
    "INPUT=unaligned_mc_tagged_polyA_filtered.bam \\\n",
    "FASTQ=unaligned_mc_tagged_polyA_filtered.fastq\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alignment - STAR\n",
    "\n",
    "*We use STAR as our RNA aligner. The manual for STAR can be found [here](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf). There are many potential\n",
    "aligners one could use at this stage, and it’s possible to substitute in your lab’s favorite. We haven’t\n",
    "tested other aligners in methodical detail, but all should produce valid BAM files that can be plugged\n",
    "into the rest of the process detailed here.\n",
    "If you’re unsure how to create an indexed reference for STAR, please read the STAR manual.\n",
    "Below is a minimal invocation of STAR. Since STAR contains a huge number of options to tailor\n",
    "alignment to a library and trade off sensitivity vs specificity, you can alter the default settings of the\n",
    "algorithm to your liking, but we find the defaults work reasonably well for Drop-seq. Be aware that\n",
    "STAR requires roughly 30 gigabytes of memory to align a single human sized genome, and 60\n",
    "gigabytes\n",
    "for our human/mouse reference.*\n",
    "\n",
    "Example:\n",
    "```bash\n",
    "/path/to/STAR/STAR \\\n",
    "--genomeDir /path/to/STAR_REFERENCE \\\n",
    "--readFilesIn unaligned_mc_tagged_polyA_filtered.fastq \\\n",
    "--outFileNamePrefix star\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SortSam\n",
    "\n",
    "*This picard program is invoked after alignment, to guarantee that the output from alignment is sorted\n",
    "in queryname order. As a side bonus, the output file is a BAM (compressed) instead of SAM\n",
    "(uncompressed.)*\n",
    "\n",
    "Example:\n",
    "\n",
    "```bash\n",
    "picard SortSam \\\n",
    "I=starAligned.out.sam \\\n",
    "O=aligned.sorted.bam \\\n",
    "SO=queryname\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MergeBamAlignment\n",
    "\n",
    "*This Picard program merges the sorted alignment output from STAR **(ALIGNED_BAM)** with the\n",
    "unaligned BAM that had been previously tagged with molecular/cell barcodes **(UNMAPPED_BAM)**.\n",
    "This recovers the BAM tags that were “lost” during alignment. The **REFERENCE_SEQUENCE** argument\n",
    "refers to the fasta metadata file.\n",
    "We ignore secondary alignments, as we want only the best alignment from STAR (or another aligner),\n",
    "instead of assigning a single sequencing read to multiple locations on the genome.*\n",
    "\n",
    "Example:\n",
    "```bash\n",
    "picard MergeBamAlignment \\\n",
    "REFERENCE_SEQUENCE=my_fasta.fasta \\\n",
    "UNMAPPED_BAM=unaligned_mc_tagged_polyA_filtered.bam \\\n",
    "ALIGNED_BAM=aligned.sorted.bam \\\n",
    "OUTPUT=merged.bam \\\n",
    "INCLUDE_SECONDARY_ALIGNMENTS=false \\\n",
    "PAIRED_RUN=false\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TagReadWithGeneExon\n",
    "\n",
    "*This is a Drop-seq program that adds a BAM tag “GE” onto reads when the read overlaps the exon of a\n",
    "gene. This tag contains the name of the gene, as reported in the annotations file. You can use either a\n",
    "GTF or a RefFlat annotation file with this program, depending on what annotation data source you\n",
    "find most useful. This is used later when we extract digital gene expression (DGE) from the BAM.*\n",
    "\n",
    "Example:\n",
    "```bash\n",
    "TagReadWithGeneExon \\\n",
    "I=merged.bam \\\n",
    "O=star_gene_exon_tagged.bam \\\n",
    "ANNOTATIONS_FILE=${refFlat} \\\n",
    "TAG=GE\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DetectBeadSynthesisErrors\n",
    "\n",
    "*In June 2015, we noticed that a recently purchased batch of ChemGenes beads generated a\n",
    "population of cell barcodes (about 10-20%) with sequences that shared the first 11 bases, but differed\n",
    "at the last base. These same cell barcodes also had a very high percentage of the base “T” at the last\n",
    "position of the UMI. Based on these observations, we concluded that a percentage of beads in the lot\n",
    "had not undergone all twelve split-and-pool bases (perhaps they had stuck to some piece of\n",
    "equipment or container, and the been re-introduced after the missing synthesis cycle). Thus, the\n",
    "20-bp Read 1 contained a mixed base at base 12 (in actuality, the first base of the UMI) and a fixed\n",
    "T-base at base 20 (in actuality, the first base of the polyT segment).*\n",
    "\n",
    "*To correct for this, we generated **DetectBeadSynthesisErrors**, which identifies cell barcodes with\n",
    "aberrant “fixed” UMI bases. If only the last UMI base is fixed as a T, the cell barcode is corrected (the\n",
    "last base is trimmed off) and all cell barcodes with identical sequence at the first 11 bases are merged\n",
    "together. If any other UMI base is fixed, the reads with that cell barcode are discarded.\n",
    "The program asks the user to select a number of barcodes on which to perform the correction. We\n",
    "use roughly 2 times the anticipated number cells, as we empirically found that this allows us to\n",
    "recover nearly every defective cell barcode that corresponds to a STAMP (rather than an empty bead\n",
    "cell barcode).*\n",
    "\n",
    "Example:\n",
    "\n",
    "```bash\n",
    "DetectBeadSynthesisErrors \\\n",
    "I=my.bam \\\n",
    "O=my_clean.bam \\\n",
    "OUTPUT_STATS=my.synthesis_stats.txt \\\n",
    "SUMMARY=my.synthesis_stats.summary.txt \\\n",
    "NUM_BARCODES= <roughly 2x the number of cells> \\\n",
    "PRIMER_SEQUENCE=AAGCAGTGGTATCAACGCAGAGTAC \\\n",
    "```\n",
    "\n",
    "*This program reads in the BAM file, and looks at the distribution of bases at each position of all UMIs\n",
    "for a cell barcode. It detects unusual distributions of base frequency, where a base with >=80%\n",
    "frequency at any position is detected as an error. Barcodes with less than 25 total UMIs are ignored.\n",
    "There are a number of different errors that are categorized:*\n",
    "\n",
    "*- **SYNTHESIS_MISSING_BASE** - 1 or more bases missing from cell barcode, resulting in fixed T’s\n",
    "at the end of UMIs. This counts the maximum number of fixed sequential T’s in the UMIs at\n",
    "the end. This error type is cleaned up by the software for situations where there is a single\n",
    "base missing, and is by far the most common error. The fix involves inserting an “N” base\n",
    "before the last cell barcode base, effectively shifting the reading frame back to where it\n",
    "should be. This will both collapse these beads back together in further analysis, as well as\n",
    "repair the UMIs for these bead barcodes.*\n",
    "\n",
    "![title](img/synthesis_missing_base.png)\n",
    "\n",
    "*- **SINGLE_UMI_ERROR** - At each position of the UMIs, the base distribution is highly skewed, i.e.\n",
    "at each position, a single base appears in >= 80% of the UMIs for that cell. There’s no fix for\n",
    "this currently. Cell barcodes with this property are dropped. These cells have the interesting\n",
    "property that the number of genes and transcripts are at a close to 1:1 ratio, as there’s\n",
    "generally only 1 UMI for every gene.*\n",
    "\n",
    "*- **PRIMER_MATCH** - Same as SINGLE_UMI_ERROR, but in addition the UMI perfectly matches\n",
    "one of the PCR primers. These cell barcodes are dropped. These errors are only detected if a\n",
    "PRIMER_SEQUENCE argument is supplied.*\n",
    "\n",
    "*- **OTHER** - UMIs are extremely skewed towards at least one base (and not T at the last base),\n",
    "but not at all 8 positions. These cell barcodes are dropped.*\n",
    "\n",
    "\n",
    "#### The file my.synthesis_stats.txt contains a bunch of useful information:\n",
    "\n",
    "*- **CELL_BARCODE** - the 12 base cell barcode*\n",
    "\n",
    "*- **NUM_UMI** - the number of total umis observed*\n",
    "\n",
    "*- **FIRST_BIASED_BASE** - the first base position where any bias is observed. ­1 for no detected\n",
    "bias*\n",
    "\n",
    "*- **SYNTH_MISSING_BASE** - as #3 but specific to runs of T’s at the end of the UMI*\n",
    "\n",
    "*- **ERROR_TYPE** - see error type definitions above*\n",
    "\n",
    "*- For bases 1-8 of the UMI, the observed base counts across all UMIs. This is a “|” delimited\n",
    "field, with counts of the A,C,G,T,N bases.*\n",
    "\n",
    "*The file my.synthesis_stats.summary.txt contains a histogram of the SYNTHESIS_MISSING_BASE\n",
    "errors, as well as the counts of all other errors, the number of total barcodes evaluated, and the\n",
    "number of barcodes ignored.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Digital Gene Expression\n",
    "\n",
    "*Extracting Digital Gene Expression (DGE) data from an aligned library is done using the Drop­seq\n",
    "program DigitalExpression. The input to this program is the aligned BAM from the alignment\n",
    "workflow. There are two outputs available: the primary is the DGE matrix, with each a row for each\n",
    "gene, and a column for each cell. The secondary analysis is a summary of the DGE matrix on a per­cell\n",
    "level, indicating the number of genes and transcripts observed.*\n",
    "\n",
    "![title](img/DGE.png)\n",
    "\n",
    "Example:\n",
    "\n",
    "```bash\n",
    "DigitalExpression \\\n",
    "I=out_gene_exon_tagged.bam \\\n",
    "O=out_gene_exon_tagged.dge.txt.gz \\\n",
    "SUMMARY=out_gene_exon_tagged.dge.summary.txt \\\n",
    "NUM_CORE_BARCODES=100\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
